File: tests/diagnostics.py
from pathlib import Path
import re,subprocess,time
project=Path(__file__).resolve().parents[1]
binary=str(project/'target/release/mono')
sections=['system','cpu','memory','storage','filesystems','temperature','network','services','boot','hardware','power','logs']
count=0
def run(args,code=0):
    global count
    start=time.monotonic()
    result=subprocess.run([binary,*args],capture_output=True,text=True,timeout=25)
    assert result.returncode==code,(args,result.stdout,result.stderr)
    count+=1
    return result.stdout,time.monotonic()-start
def validate(text,size):
    states=re.findall(r'^(System|CPU|Memory|Storage|Filesystems|Temperature|Network|Services|Boot|Hardware|Power|Logs)\s+(OK|WARNING|ERROR|UNKNOWN|SKIPPED)$',text,re.M)
    assert len(states)==size,(size,states,text)
    problems=sum(status in ['WARNING','ERROR'] for _,status in states)
    assert f'Problems found: {problems}' in text
    assert len(text.splitlines())<110
text,elapsed=run(['diagnose']);validate(text,12)
for section in sections:
    text,_=run(['diagnose',section]);validate(text,12 if section=='system' else 1)
run(['diagnose','invalid'],1);run(['diagnose','memory','extra'],1)
text,_=run(['diagnose','system','--dry-run']);assert 'nothing executed' in text and 'TCP/443' in text
run(['diagnose','bad','--dry-run'],1)
for args in [['help','diagnose'],['manual','diagnose'],['manual','backend','diagnose']]:run(args)
print(f'{count} diagnostics CLI checks passed; system report completed in {elapsed:.2f}s.')

File: crates/mono-cli/src/diagnostics.rs
use mono_core::{
    Result,
    diagnostics::{Health, SECTIONS},
};

pub fn run(section: &str, dry_run: bool) -> Result<()> {
    if !SECTIONS.contains(&section) {
        return Err(format!(
            "Unknown section: {section}. Run mono help diagnose"
        ));
    }
    if dry_run {
        println!("Diagnostics: {section}. Read-only checks; nothing executed.");
        if matches!(section, "system" | "network") {
            println!("Network probes: DNS archlinux.org; TCP/443 to 1.1.1.1 or 9.9.9.9.");
        }
        return Ok(());
    }
    let checks = mono_platform_arch::diagnostics::diagnose(section)?;
    println!("Monux System Diagnostics\n");
    for c in &checks {
        println!("{:<13} {}", c.name, c.health.label());
    }
    let problems = checks.iter().filter(|c| c.health.problem()).count();
    let unknown = checks
        .iter()
        .filter(|c| c.health == Health::Unknown)
        .count();
    let skipped = checks
        .iter()
        .filter(|c| c.health == Health::Skipped)
        .count();
    println!("\nProblems found: {problems}");
    if unknown + skipped > 0 {
        println!("Unavailable: {unknown}; skipped: {skipped}");
    }
    for c in &checks {
        if c.health != Health::Ok || section != "system" {
            let mark = match c.health {
                Health::Error => "X",
                Health::Warning => "!",
                Health::Ok => "+",
                _ => "?",
            };
            println!("\n[{mark}] {}", c.name);
            for line in &c.details {
                println!("    {line}");
            }
        }
    }
    if problems > 0 {
        println!("\nRun:");
        for c in checks.iter().filter(|c| c.health.problem()) {
            println!("    mono diagnose {}", c.name.to_ascii_lowercase());
        }
    }
    // Diagnostic findings are report data. Invocation failures still use the CLI error path.
    Ok(())
}
